### PR TITLE
spirv-fuzz: Replace dead-block terminators with OpKill etc.

### LINKED ISSUE #3615: spirv-fuzz: Add dead OpKill, OpReturn, OpUnreachable
Introduce a new transformation that an change the terminator of a dead block to OpKill, OpReturn/OpReturnValue or OpUnreachable.  This is only legal under certain conditions, and I expect it might be best to check some obvious conditions in IsApplicable, and then do the trick of cloning the module, applying the change and invoking the validator to do a full check.

We already have a transformation to add a dead block, so we could use this transformation to build a fuzzer pass that finds places to add a dead block and then immediately change its terminator to one of these early-exit instructions.

OpKill should only be used in a fragment shader.

### PR DESCRIPTION
**Pull request by afd**
Fixes #3615.

### CODE REVIEW

File: source/fuzz/fuzzer_pass_replace_branches_from_dead_blocks_with_exits.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    ~FuzzerPassReplaceBranchesFromDeadBlocksWithExits() = default;
+
+void FuzzerPassReplaceBranchesFromDeadBlocksWithExits::Apply() {
+  // OpKill can only be used as a terminator in a function that is guaranteed
+  // to be executed with the Fragment execution model.  We conservatively only
+  // allow OpKill if every entry point in the module has the Fragment execution
+  // model.
+  auto fragment_execution_model_guaranteed =
+      std::all_of(GetIRContext()->module()->entry_points().begin(),
+                  GetIRContext()->module()->entry_points().begin(),
```
Comment (Vasniktel) [inline source/fuzz/fuzzer_pass_replace_branches_from_dead_blocks_with_exits.cpp:45]: `GetIRContext()->module()->entry_points().begin()` -> `GetIRContext()->module()->entry_points().end()`.

**Review by Vasniktel — CHANGES_REQUESTED** [review]

File: source/fuzz/fuzzer_pass_replace_branches_from_dead_blocks_with_exits.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_replace_branches_from_dead_blocks_with_exits.cpp]: Fixed, thanks!

File: source/fuzz/transformation_replace_branch_from_dead_block_with_exit.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  if (message_.opcode() == SpvOpReturnValue) {
+    new_operands.push_back(
+        {SPV_OPERAND_TYPE_TYPE_ID, {message_.return_value_id()}});
+  }
+  auto terminator =
+      ir_context->get_instr_block(message_.block_id())->terminator();
+  terminator->SetOpcode(static_cast<SpvOp>(message_.opcode()));
+  terminator->SetInOperands(std::move(new_operands));
+  ir_context->InvalidateAnalysesExceptFor(opt::IRContext::kAnalysisNone);
+}
```
Comment (paulthomson) [inline source/fuzz/transformation_replace_branch_from_dead_block_with_exit.cpp:106]: Do you also need to potentially update OpPhi instructions in the successor block?

**Review by Vasniktel — APPROVED** [review]

File: source/fuzz/transformation_replace_branch_from_dead_block_with_exit.cpp (hunk 2 above)
Comment (afd) [reply source/fuzz/transformation_replace_branch_from_dead_block_with_exit.cpp]: You are right - done.

**Pull request merged**